src/lib/server/postgres_protocol_handler.cpp

```diff
@@ -0,0 +1,265 @@
+#include "postgres_protocol_handler.hpp"
+
+namespace opossum {
+
+template <typename SocketType>
+PostgresProtocolHandler<SocketType>::PostgresProtocolHandler(const std::shared_ptr<SocketType>& socket)
+    : _read_buffer(socket), _write_buffer(socket) {}
+
+template <typename SocketType>
+uint32_t PostgresProtocolHandler<SocketType>::read_startup_packet_header() {
+  // Special SSL version number that we catch to deny SSL support
+  constexpr auto ssl_request_code = 80877103u;
+
+  const auto body_length = _read_buffer.template get_value<uint32_t>();
+  const auto protocol_version = _read_buffer.template get_value<uint32_t>();
+
+  // We currently do not support SSL
+  if (protocol_version == ssl_request_code) {
+    _ssl_deny();
+    return read_startup_packet_header();
+  } else {
+    // Substract uint32_t twice, since both packet length and protocol version have been read already
+    return body_length - 2 * LENGTH_FIELD_SIZE;
+  }
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::read_startup_packet_body(const uint32_t size) {
+  // As of now, we don't do anything with the startup packet body. It contains authentication data and the
+  // database name the user desires to connect to. Hence, we read this information from the network device as
+  // one large string and throw it away.
+  _read_buffer.get_string(size, HasNullTerminator::No);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_authentication_response() {
+  _write_buffer.template put_value(PostgresMessageType::AuthenticationRequest);
+  // Since we don't have any authentication mechanism, authentication is always successful
+  constexpr uint32_t authentication_successful = 0;
+  _write_buffer.template put_value<uint32_t>(sizeof(LENGTH_FIELD_SIZE) + sizeof(authentication_successful));
+  _write_buffer.template put_value<uint32_t>(authentication_successful);
+}
+
+template <typename SocketType>
+void PostgresProtocolHandler<SocketType>::send_parameter(const std::string& key, const std::string& value) {
+  const auto packet_size = sizeof(LENGTH_FIELD_SIZE) + key.size() + value.size() + 2u /* null terminator */;
```

Comment:
Same question as @mrks asked below: Why +2 and not +1? Are those two null terminators for the two strings?

Reply:
Yes, added a comment for this.